ccl/sqlproxyccl/tenant: TestWatchTenants failed

ccl/sqlproxyccl/tenant.TestWatchTenants [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10122787?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10122787?buildTab=artifacts#/) on master @ [76078816b69ee6d1f8dacd581ef464660fb1b697](https://github.com/cockroachdb/cockroach/commits/76078816b69ee6d1f8dacd581ef464660fb1b697):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x4e2bd31]

```
Stack: 

```
goroutine 16050 [running]:
testing.tRunner.func1.2({0x56a4da0, 0xa7138e0})
	GOROOT/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1399 +0x39f
panic({0x56a4da0, 0xa7138e0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:122 +0x2a9
panic({0x56a4da0, 0xa7138e0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc00017f398?, {0x75a3c70, 0xc0000c6058})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x56a4da0, 0xa7138e0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc000dff080, {0x75a3c70, 0xc0000c6058})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:507 +0x21d
panic({0x56a4da0, 0xa7138e0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test.TestWatchTenants(0xc002448340)
	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:221 +0xed1
testing.tRunner(0xc002448340, 0x63403a0)
	GOROOT/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1493 +0x35f
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestWatchTenants
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestWatchTenants985868698
    test_log_scope.go:79: use -show-logs to present logs inline
*
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:507
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | runtime.panicmem
*   | 	GOROOT/src/runtime/panic.go:260
*   | runtime.sigpanic
*   | 	GOROOT/src/runtime/signal_unix.go:835
*   | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test.TestWatchTenants
*   | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:221
*   | testing.tRunner
*   | 	GOROOT/src/testing/testing.go:1446
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_amd64.s:1594
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
*
    stopper.go:229: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestWatchTenants985868698
    stopper.go:229: panic: runtime error: invalid memory address or nil pointer dereference
--- FAIL: TestWatchTenants (0.08s)
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchTenants.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-28029

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -414,6 +414,7 @@ func TestPrivateEndpointsACL(t *testing.T) {
 					time.Second, 5*time.Millisecond,
 					"Expected the connection to eventually fail",
 				)
+				require.Error(t, err)
 				require.Regexp(t, "connection reset by peer|unexpected EOF", err.Error())
 				require.Equal(t, int64(1), s.metrics.ExpiredClientConnCount.Count())
 			},
@@ -1203,6 +1204,7 @@ func TestDenylistUpdate(t *testing.T) {
 		time.Second, 5*time.Millisecond,
 		"Expected the connection to eventually fail",
 	)
+	require.Error(t, err)
 	require.Regexp(t, "closed|bad connection", err.Error())
 	require.Equal(t, int64(1), s.metrics.ExpiredClientConnCount.Count())
 }

--- a/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go
+++ b/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go
@@ -204,12 +204,13 @@ func TestWatchTenants(t *testing.T) {
 		if tds.WatchTenantsListenersCount() != 0 {
 			return errors.New("watchers have not been removed yet")
 		}
+
+		// Make sure watcher has attempted to restart, and invalidates entries.
+		if _, err := dir.LookupTenant(ctx, tenant10); err == nil {
+			return errors.New("entries have not been invalidated")
+		}
 		return nil
 	})
-
-	// Entry was invalidated.
-	_, err = dir.LookupTenant(ctx, tenant10)
-	require.Regexp(t, "directory server has not been started", err.Error())
 
 	// Trigger events, which will be missed by the tenant watcher.
 	tds.DeleteTenant(tenant10)

